File: src/main/java/com/IceCreamQAQ/Yu/event/EventBus.java
package com.IceCreamQAQ.Yu.event;

import com.IceCreamQAQ.Yu.event.events.Event;
import lombok.extern.slf4j.Slf4j;
import lombok.val;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.inject.Inject;
import java.lang.reflect.Parameter;
import java.util.ArrayList;
import java.util.List;

@Slf4j
public class EventBus {

    @Inject
    private EventInvokerCreator creator;

    private List<EventListenerInfo>[] eventInvokersLists;

    public EventBus() {
        eventInvokersLists = new List[3];
        for (int i = 0; i < 3; i++) {
            eventInvokersLists[i] = new ArrayList<>();
        }
    }

    public boolean post(Event event) {
        for (val list : eventInvokersLists) {
            for (val eventInvoker : list) {
                try {
                    eventInvoker.getInvoker().invoke(event);
                } catch (Throwable throwable) {
                    val errorMessageBuilder = new StringBuilder("EventListenerError! At: ");
                    errorMessageBuilder.append(eventInvoker.getClazz().getName()).append(".").append(eventInvoker.getMethod().getName()).append("(");
                    if (eventInvoker.getMethod().getParameters().length != 0) {
                        val max = eventInvoker.getMethod().getParameters().length - 1;
                        for (int i = 0; i < eventInvoker.getMethod().getParameters().length; i++) {
                            val para = eventInvoker.getMethod().getParameters()[i];
                            errorMessageBuilder.append(para.getType().getSimpleName());
                            if (i != max) errorMessageBuilder.append(", ");
                        }
                    }
                    errorMessageBuilder.append(")");
                    log.error(errorMessageBuilder.toString(), throwable);
                }
                if (event.cancelAble() && event.cancel) return true;
            }
        }
        return false;
    }

    public void register(Object object) {
        val eventInvokersLists = creator.register(object);
        for (int i = 0; i < 3; i++) {
            this.eventInvokersLists[i].addAll(eventInvokersLists[i]);
        }
    }

}

File: src/main/kotlin/com/IceCreamQAQ/Yu/event/EventBus.kt
package com.IceCreamQAQ.Yu.event

import com.IceCreamQAQ.Yu.annotation.AutoBind
import com.IceCreamQAQ.Yu.event.events.Event
import com.IceCreamQAQ.Yu.annotation.Event.Weight
import com.IceCreamQAQ.Yu.event.events.CancelAbleEvent
import com.IceCreamQAQ.Yu.event.events.EventListenerRunExceptionEvent
import org.slf4j.LoggerFactory
import java.util.*

@AutoBind
interface EventBus {
    fun post(event: Event): Boolean
    fun register(info: EventListenerInfo)
    fun unregister(info: EventListenerInfo)
}

class EventBusImpl : EventBus {

    companion object {
        private val log = LoggerFactory.getLogger(EventBus::class.java)
    }

    private val eis: Map<Weight, MutableList<EventListenerInfo>> =
        hashMapOf(
            Weight.record to arrayListOf(),
            Weight.highest to arrayListOf(),
            Weight.high to arrayListOf(),
            Weight.normal to arrayListOf(),
            Weight.low to arrayListOf(),
            Weight.lowest to arrayListOf(),
        )

    val ss = arrayOf(Weight.highest, Weight.high, Weight.normal, Weight.low, Weight.lowest)

    override fun post(event: Event): Boolean {
        val cancelable = event is CancelAbleEvent
        operator fun EventListenerInfo.invoke(): Boolean {
            try {
                invoker.invoke(event)
            } catch (throwable: Throwable) {
                log.error("EventListenerError! At: $methodFullName.", throwable)
                if (event !is EventListenerRunExceptionEvent)
                    post(EventListenerRunExceptionEvent(this, throwable))
            }
            return cancelable && (event as CancelAbleEvent).isCanceled
        }
        eis[Weight.record]!!.forEach { it() }
        if (cancelable)(event as CancelAbleEvent).isCanceled = false
        for (i in 0..4) {
            val width = ss[i]
            for (eventInvoker in eis[width]!!) {
                if (eventInvoker()) return true
            }
        }
        return false
    }

    override fun register(info: EventListenerInfo) {
        eis[info.weight]!!.add(info)
    }

    override fun unregister(info: EventListenerInfo) {
        eis[info.weight]!!.remove(info)
    }

}
